Add visit all capability

migueleliasweb opened the pull request.

migueleliasweb committed:
Add visit all capability to tree and nodes

migueleliasweb commented:
Fixes: https://github.com/xlab/treeprint/issues/9

elliotchance commented on treeprint.go:
```diff
@@ -159,6 +169,24 @@ func (n *node) SetMetaValue(meta MetaValue) {
 
+func (n *node) VisitAll(fn NodeVisitor) {
+	for _, node := range n.Nodes {
+		// if len(node.Nodes) == 0 {
```
Left over debugging code?

migueleliasweb replied:
Well, not really. My idea here was to give an example on how to determine if the item you're iterating is a branch or a node. Should I take the comments out and add them as normal text in the docs?

migueleliasweb replied:
Scratch all that. I was looking at the wrong place. Yeah it was just me testing. :facepalm: . Removed it.

elliotchance commented on treeprint.go:
```diff
@@ -40,6 +46,10 @@ type Tree interface {
+
+	// VisitAll iterates over the tree, branches and nodes.
+	// If need to iterate over the whole tree, use the root node.
+	VisitAll(fn NodeVisitor)
```
Might be nice to mention here that this is a depth-first walk.

migueleliasweb replied:
I was never the good regarding algorithm names but since the code is iterating over all items in the current branch first to them head to the next branch, wouldn't this be breadth first? Btw, yeah it could be worth adding that info to the docs, you're right 👍.

elliotchance reviewed (commented):
This is not my repo so I won't require changes.

xlab commented:
Hi! Thanks for providing both PRs. I've merged the first one, as it was very lightweight. This one I guess has reasonable comments from @elliotchance , would you mind adding those changes? :)

migueleliasweb commented:
> Hi! Thanks for providing both PRs. I've merged the first one, as it was very lightweight. This one I guess has reasonable comments from @elliotchance , would you mind adding those changes? :)

I will have a go I'm a few hours! 🤟

migueleliasweb commented:
I just rebased this branch so it's nice and easy to merge, removed the debugging code and added a comment regarding the approach taken for the `VisitAll`. Could you have a look again, @elliotchance ?
